File: 112-path-sum/112-path-sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} targetSum
 * @return {boolean}
 */
var hasPathSum = function(root, targetSum) {
    return dfs(root, targetSum, 0);
};

function dfs(node, target, acc) {
    if (!node) {
        return false;
    }
    acc += node.val;
    if(node.left == null && node.right == null){
        return acc == target;
    }
    return dfs(node.left, target, acc) || dfs(node.right, target, acc);
}

File: 112-path-sum/112-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        return dfs(root, targetSum, 0);
    }
    
    bool dfs(TreeNode* node, int target, int acc) {
        if (!node) {
            return false;
        }
        acc += node->val;
        if(node->left == nullptr && node->right == nullptr){
            return acc == target;
        }
        return dfs(node->left, target, acc) || dfs(node->right, target, acc);
    }
};